### PR TITLE
Pyg to hls rebase

### PR DESCRIPTION
**Pull request by abdelabd**
Backend:
- added nnet_graph.h, has the functions for the main EdgeBlock/NodeBlock architecture, as well as the Aggregate function
- added onto nnet_array.h, two functions to convert between 1D and 2D arrays
- added onto nnet_activation.h (sigmoid_batch)

Frontend:
- hls_model.py, new class: HLSModel_GNN(HLSModel)
- hls_layers.py, new classes: Aggregate(Layer), GraphBlock(Layer), EdgeBlock(Layer), NodeBlock(Layer)
- vivado_writer.py, new class: VivadoWriter_GNN(VivadoWriter) <-- overrides self.write_yml() and self.write_testbench()
- hls4ml/converters/pyg_to_hls.py: defines class PygModelReader(PyTorchModelReader), and the top-level conversion function pyg_to_hls().

### CODE REVIEW

File: hls4ml/converters/pyg_to_hls.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+        # get inputs, outputs
+        index = len(layer_list)+1
+        if val=="NodeBlock":
+            layer_dict["inputs"] = [last_node_update, last_edge_aggr_update]
+            layer_dict["outputs"] = [f"layer{index}_out_P"]
+            last_node_update = f"layer{index}_out_P"
+        elif val=="EdgeBlock":
+            layer_dict["inputs"] = [last_node_update, last_edge_update, "edge_index"]
+            layer_dict["outputs"] = [f"layer{index}_out_L", f"layer{index}_out_Q"]
```
Comment (jmduarte) [inline hls4ml/converters/pyg_to_hls.py]: should these be renamed `_out` and `out_aggr` ?

File: hls4ml/templates/vivado/nnet_utils/nnet_activation.h [hunk 2]
```diff
@@ -171,6 +188,23 @@ void  sigmoid(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
     }
 }
 
+template<class data_T, class res_T, typename CONFIG_T>
+void  sigmoid_batch(data_T data[CONFIG_T::n_batch][CONFIG_T::n_in], res_T res[CONFIG_T::n_batch][CONFIG_T::n_in])
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_activation.h]: if we no longer use `sigmoid_batch`, we could remove it

File: hls4ml/templates/vivado/nnet_utils/nnet_activation.h [hunk 3]
```diff
@@ -477,6 +511,23 @@ void  tanh(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
     }
 }
 
+template<class data_T, class res_T, typename CONFIG_T>
+void  tanh_batch(data_T data[CONFIG_T::n_batch][CONFIG_T::n_in], res_T res[CONFIG_T::n_batch][CONFIG_T::n_in])
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_activation.h]: if we no longer use `tanh_batch`, we could remove it

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+
+#include "nnet_common.h"
+#include "nnet_merge.h"
+#include "nnet_dense.h"
+#include "nnet_dense_resource.h"
+#include "nnet_activation.h"
+#include "nnet_array.h"
+#include "hls_stream.h"
+#include <string>
+#include <sstream>
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: do we need this still?

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    // Layer Sizes
+    static const unsigned n_node = 10;
+    static const unsigned n_edge = 20;
+    static const unsigned n_features = 3;
+    static const unsigned e_features = 4;
+    static const unsigned n_out = 4;
+    static const unsigned n_layers = 3;
+
+    // message-passing parameters
+    static const unsigned aggr = 0;
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: can we use an enum?

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+    static const unsigned flow = 0;
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: can we use an enum?

File: hls4ml/writer/vivado_writer.py [hunk 7]
```diff
@@ -654,3 +656,184 @@ def write_hls(self, model):
         self.write_yml(model)
         self.write_tar(model)
         print('Done')
+ 
+class VivadoWriter_GNN(VivadoWriter):
+
+    def write_test_bench(self, model):
```
Comment (jmduarte) [inline hls4ml/writer/vivado_writer.py]: this function can be removed... see: https://github.com/abdelabd/hls4ml/pull/6

File: hls4ml/templates/vivado/myproject_gnn_test.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
+#include <fstream>
```
Comment (jmduarte) [inline hls4ml/templates/vivado/myproject_gnn_test.cpp]: this file can be removed... see https://github.com/abdelabd/hls4ml/pull/6

File: hls4ml/writer/vivado_writer.py [hunk 9]
```diff
@@ -654,3 +656,184 @@ def write_hls(self, model):
...
+                newline = line
+                for out in model.get_output_variables():
+                    newline += indent + 'nnet::print_result<{}, {}>({}, std::cout, true);\n'.format(out.type.name, out.size_cpp(), out.cppname)
+            else:
+                newline = line
+            fout.write(newline)
+        f.close()
+        fout.close()
+
+    def write_yml(self, model):
```
Comment (jmduarte) [inline hls4ml/writer/vivado_writer.py]: can we generalize the existing `write_yml` to work for this model?

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+    if(CONFIG_T::io_stream){
+      #pragma HLS STREAM variable=X
+    }
+    #pragma HLS PIPELINE II=CONFIG_T::reuse_factor
+    data_T R0_logits[CONFIG_T::dense_config1::n_batch][CONFIG_T::dense_config1::n_out];
+    #pragma HLS ARRAY_PARTITION variable=R0_logits complete dim=0
+    nnet::dense_batch<data_T, data_T, typename CONFIG_T::dense_config1>(X, R0_logits, w0, b0);
+    data_T R0[CONFIG_T::relu_config1::n_batch][CONFIG_T::relu_config1::n_in];
+    #pragma HLS ARRAY_PARTITION variable=R0 complete dim=0
+    nnet::relu_batch<data_T, data_T, typename CONFIG_T::relu_config1>(R0_logits, R0);
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: it looks like this is the only place `relu_batch` is used... but maybe we should unroll the array and use `relu` instead?

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 11]
```diff
@@ -0,0 +1,564 @@
...
+
+    if(CONFIG_T::activate_final){
+        data_T R_logits[CONFIG_T::dense_config2::n_batch][CONFIG_T::dense_config2::n_out];
+        #pragma HLS ARRAY_PARTITION variable=R_logits complete dim=0
+        nnet::dense_batch<data_T, data_T, typename CONFIG_T::dense_config2>(R0, R_logits, w1, b1);
+        nnet::relu_batch<data_T, res_T, typename CONFIG_T::relu_config2>(R_logits, R);
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: ditto about `relu_batch`

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 12]
```diff
@@ -0,0 +1,564 @@
...
+      }
+    }
+
+    // output array --> output vector
+    nnet::mat_to_vec<res_T, res_T, typename CONFIG_T::node_update_config>(node_update, node_update_1D);
+
+  }
+
+  template<class data_T, class res_T, typename CONFIG_T>
+    void graph_independent(
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: we could potentially remove this (unless we add a GNN example that uses it)

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 13]
```diff
@@ -0,0 +1,582 @@
...
+    // message-passing parameters
+    static const unsigned aggr = aggr_sum;
+    static const unsigned flow = source_to_target;
+    
+    // Resource reuse info
+    static const unsigned io_type = io_parallel;
+    static const unsigned reuse_factor = 1;
+    static const unsigned n_zeros = 0;
+
+    static const bool activate_final = false;
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: should this really be a boolean? I saw `activate_final` set to `sigmoid` in other places. Is this set anywhere?

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h [hunk 14]
```diff
@@ -0,0 +1,582 @@
...
+    )
+  {
+    min_rep = -1;
+    for(int i=0; i<fp_int_bits-1; i++){
+      min_rep = min_rep*2;
+    }
+  }
+
+  template<class data_T, class res_T, typename CONFIG_T>
+    void dense_mult_1lyr(
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_graph.h:95]: lower priority, but i will try to consolidate this 1lyr, 2lyr, 3lyr, 4lyr code

File: hls4ml/templates/vivado/nnet_utils/nnet_dense_resource.h [hunk 15]
```diff
@@ -279,6 +279,282 @@ void dense_resource(
...
+    	}
+        dense_resource<data_T, res_T, CONFIG_T>(data_temp, res_temp, weights, biases);
+        for (int ii = 0; ii < CONFIG_T::n_out; ii++) {
+          res[bb][ii] = res_temp[ii];
+        }
+    }
+}
+
+template<class data_T, class res_T, typename CONFIG_T>
+void dense_resource_rf_leq_nin_basic(
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_dense_resource.h]: i think we can get rid of this, but i will submit a PR to do it

File: hls4ml/writer/vivado_writer.py [hunk 16]
```diff
@@ -305,6 +305,8 @@ def write_defines(self, model):
                 newline = line
                 numbers = OrderedDict.fromkeys([layer.get_numbers_cpp() for layer in model.get_layers()])
                 newline += ''.join(numbers)
+                newline = '\n'.join(set(newline.split('\n')))
```
Comment (jmduarte) [inline hls4ml/writer/vivado_writer.py]: try `newline += ''.join(set(numbers))` on the previous line instead (or something like that)

File: hls4ml/model/hls_layers.py [hunk 17]
```diff
@@ -1810,6 +1814,614 @@ def _get_transforms_config(self, params):
...
+
+        for name, module in self.submodules.items():
+            if module.__class__.__name__ == 'Linear':
+                data = self.model.get_weights_data(self.name, name, 'kernel').transpose()
+                var_name = f"{self.name}_w{linear_count}"
+                self.add_weights_variable(name=var_name, var_name=var_name, data=data, quantizer=quantizer,
+                                              compression=compression)
+                linear_count += 1
+
+        # DUMMIES
```
Comment (jmduarte) [inline hls4ml/model/hls_layers.py:1947]: Why are these weights added twice with different names?

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: hls4ml/model/hls_layers.py [hunk 18]
```diff
@@ -1810,6 +1814,614 @@ def _get_transforms_config(self, params):
...
+        params['n_node'] = self.n_node_cppname
+        params['n_edge'] = self.n_edge_cppname
+        params['node_dim'] = self.node_dim_cppname
+        params['edge_dim'] = self.edge_dim_cppname
+        params['out_dim'] = self.out_dim
+        params['n_layers'] = self.attributes["n_layers"]
+        params['io_type'] = 'io_parallel'
+        params['reuse'] = 1
+        params['n_zeros'] = 0
+        params['fp_int_bits'] = self.fp_type.integer
```
Comment (jmduarte) [inline hls4ml/model/hls_layers.py]: I think we can get the integer part directly from HLS... I will make a PR

File: hls4ml/templates/vivado/nnet_utils/nnet_dense_resource.h [hunk 19]
```diff
@@ -279,6 +284,27 @@ void dense_resource(
     }
 }
 
+template<class data_T, class res_T, typename CONFIG_T>
+void dense_batch(
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_dense_resource.h]: remove `dense_batch`

File: hls4ml/converters/pyg_to_hls.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+        return data
+
+def pyg_to_hls(model, forward_dict, graph_dims,
+               activate_final = None,
+               fixed_precision_bits=16,
+               fixed_precision_int_bits=6,
+               int_precision_bits=16,
+               int_precision_signed=False,
+               output_dir = None,
+               reuse = 1):
```
Comment (jmduarte) [inline hls4ml/converters/pyg_to_hls.py]: We should pass in a single `HLSConfig` object instead of each of these explicit arguments. See how it's done with `keras_to_hls`:
https://github.com/fastmachinelearning/hls4ml/blob/master/hls4ml/converters/keras_to_hls.py#L193
or `pytorch_to_hls`:
https://github.com/Duchstf/hls4ml/blob/onnx_dev_pytorch/hls4ml/converters/pytorch_to_hls.py#L106

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: hls4ml/model/hls_layers.py (hunk 17 above)
Comment (abdelabd) [reply hls4ml/model/hls_layers.py:1947]: Are you reffering to the "add_weight" vs "add_bias" functions?

Comment (jmduarte) [reply hls4ml/model/hls_layers.py:1947]: I'm referring to these two lines:

https://github.com/Duchstf/hls4ml/pull/4/files/0318f2843be7f9c6a9009c57ffcb134ecd41aec1#diff-596f42295045305371ca3a443fce3a9c78302f2462ded2f2b16e2c3791944e95R1825

https://github.com/Duchstf/hls4ml/pull/4/files/0318f2843be7f9c6a9009c57ffcb134ecd41aec1#diff-596f42295045305371ca3a443fce3a9c78302f2462ded2f2b16e2c3791944e95R1832

File: hls4ml/templates/vivado/nnet_utils/nnet_graph.h (hunk 13 above)
Comment (abdelabd) [reply hls4ml/templates/vivado/nnet_utils/nnet_graph.h]: This is from Vesal's original code. I think you're thinking of "final_act" but that is only in the front-end afaik... you're right though, I should probably update these naming conventions

File: .idea/hls4ml.iml [hunk 21]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmduarte) [inline .idea/hls4ml.iml]: remove file

File: .idea/hls4ml.iml [hunk 22]
```diff
@@ -0,0 +1,8 @@
...
+<module type="PYTHON_MODULE" version="4">
```
Comment (jmduarte) [inline .idea/hls4ml.iml]: remove file

File: .idea/inspectionProfiles/profiles_settings.xml [hunk 23]
```diff
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
```
Comment (jmduarte) [inline .idea/inspectionProfiles/profiles_settings.xml]: remove file

File: .idea/modules.xml [hunk 24]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmduarte) [inline .idea/modules.xml]: remove file

File: hls4ml/.idea/misc.xml [hunk 25]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmduarte) [inline hls4ml/.idea/misc.xml]: remove file

File: hls4ml/.idea/modules.xml [hunk 26]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmduarte) [inline hls4ml/.idea/modules.xml]: remove file

File: hls4ml/.idea/workspace.xml [hunk 27]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmduarte) [inline hls4ml/.idea/workspace.xml]: remove file

File: hls4ml/converters/pyg_to_hls.py [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+
+    # get graph dimensions
+    n = graph_dims.get("n_node", 112)
+    m = graph_dims.get("n_edge", 148)
+    p = graph_dims.get("node_dim", 3)
+    q = graph_dims.get("edge_dim", 4)
+    r = graph_dims.get("relation_dim", q)
```
Comment (jmduarte) [inline hls4ml/converters/pyg_to_hls.py]: What's the difference between `edge_dim` and `relation_dim` again? Do we need both?

File: hls4ml/templates/vivado/build_prj.tcl [hunk 29]
```diff
@@ -5,10 +5,10 @@ array set opt {
   reset      0
   csim       1
   synth      1
-  cosim      1
-  validation 1
+  cosim      0
+  validation 0
   export     0
-  vsynth     0
+  vsynth     1
```
Comment (jmduarte) [inline hls4ml/templates/vivado/build_prj.tcl]: reset to previous defaults

File: hls4ml/templates/vivado/nnet_utils/nnet_activation.h [hunk 30]
```diff
@@ -114,6 +114,23 @@ void  relu1(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
     relu_max<data_T, res_T, 1, CONFIG_T>(data, res);
 }
 
+template<class data_T, class res_T, typename CONFIG_T>
+void  relu_batch(data_T data[CONFIG_T::n_batch][CONFIG_T::n_in], res_T res[CONFIG_T::n_batch][CONFIG_T::n_in])
```
Comment (jmduarte) [inline hls4ml/templates/vivado/nnet_utils/nnet_activation.h]: remove `relu_batch`

**Review by jmduarte — CHANGES_REQUESTED** [review]
